examples/channelscan/main.go

```diff
@@ -0,0 +1,71 @@
+// This example program shows using Go routines and channels to coordinate
+// BLE scanning.
+//
+// The first Go routine starts scanning using the BLE adaptor. When it finds
+// a new device, it puts the information into a channel so it can be displayed.
+//
+// The second Go routine is a ticker that puts a "true" value into a channel every 3 seconds.
+//
+// The main function uses a select{} statement to wait until one of the two channels is unblocked
+// by receiving data. If a new device is found, the boolean variable named "found" will
+// be set to true, so that the timeout is reset for each 3 second period.
+package main
+
+import (
+	"time"
+
+	"tinygo.org/x/bluetooth"
+)
+
+var (
+	adapter = bluetooth.DefaultAdapter
+	devices = make(chan bluetooth.ScanResult, 1)
+	ticker  = make(chan bool, 1)
+	found   = true
+)
+
+func main() {
+	// Enable BLE interface.
+	if err := adapter.Enable(); err != nil {
+		panic("failed to enable adaptor:" + err.Error())
+	}
+
+	// Start scanning
+	go performScan()
+
+	// Start timeout ticker
+	go startTicker()
+
+	// Wait for devices to be scanned
+	for {
+		select {
+		case device := <-devices:
+			found = true
+			println("found device:", device.Address.String(), device.RSSI, device.LocalName())
+		case <-ticker:
+			if !found {
+				println("no devices found in last 3 seconds...")
+			}
+			found = false
+		}
+	}
+}
+
+func performScan() {
+	println("scanning...")
+
+	err := adapter.Scan(func(adapter *bluetooth.Adapter, device bluetooth.ScanResult) {
+		devices <- device
```

Comment:
There is actually still a short race condition here.
Basically, the scan result will get overwritten once the next scan result comes in. So if the main goroutine isn't fast enough to process this scan result, it might display some duplicate scan results (and miss one of them).

I'm not sure how (or if) this should be fixed. One possibly expensive way is to stop scanning on each incoming packet, and start it again after the result is displayed.

Reply:
One solution would be to create local struct with fields we need from `bluetooth.ScanResult` and make the channel of that type
```suggestion
		devices <- scanData{
			name:    device.LocalName(),
			address: device.Address.String(),
			rssi:    device.RSSI,
		}
```

Reply:
No. Both `device.LocalName()` and `device.Address.String()` do heap allocations which are not allowed inside interrupts.

Reply:
Actually, looking again at the code, the issue is bigger: `devices <- device` is a blocking send. This will usually work if the main goroutine is fast enough, but there is no guarantee that it will be. It would be better to have a non-blocking send like this:

```go
select {
case devices <- device:
default:
}
```

Reply:
@aykevl I actually ran my suggestion on NRF52 board and didn't get panic even if `-print-allocs` shows heap allocation. Looking at the `Scan` implementation for NRF52, the provided callback is called directly (from the context of `Scan`) and not from interrupt handler. Also there is a comment that scan is stopped when first result comes and a call to restart the scan every time the result comes. If this is correct (I would have to check SD documentation) it would mean that on NRF52 device there would be no race (with my suggestion) and also that blocking the `Scan` callback would not be an issue since scan is stopped during it's execution and restarted after (which explains why I get so many duplicate results in the scan). I didn't look at implementations for other targets but I hope the scan callback is also decoupled there because it is much more convenient that way.

Reply:
@HattoriHanzo031 I checked the code and you are correct. The code in the PR seems fine.